toolchain/base/int_store.h

```diff
@@ -0,0 +1,422 @@
+    if (IntId id = TryMakeSignedValue(value); id.is_valid()) [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return LookupSignedLarge(std::move(value));
+  }
+
+  // Output a YAML description of this data structure. Note that this will only
+  // include the integers that required storing, not those successfully embedded
+  // into the ID space.
+  auto OutputYaml() const -> Yaml::OutputMapping;
+
+  auto array_ref() const -> llvm::ArrayRef<llvm::APInt> {
+    return values_.array_ref();
+  }
+  auto size() const -> size_t { return values_.size(); }
+
+  // Collects the memory usage of the separately stored integers.
+  auto CollectMemUsage(MemUsage& mem_usage, llvm::StringRef label) const
+      -> void;
+
+ private:
+  friend struct Testing::IntStoreTestPeer;
+
+  struct APIntId : IdBase, Printable<APIntId> {
+    using ValueType = llvm::APInt;
+    static const APIntId Invalid;
+    using IdBase::IdBase;
+    auto Print(llvm::raw_ostream& out) const -> void {
+      out << "ap_int";
+      IdBase::Print(out);
+    }
+  };
+
+  static constexpr int MinAPWidth = 64;
+
+  static auto MakeIndexOrInvalid(int index) -> IntId {
+    CARBON_DCHECK(index >= 0 && index <= IntId::InvalidIndex);
+    return IntId(IntId::ZeroIndexId - index);
+  }
+
+  // Tries to make a signed 64-bit integer into an embedded value in the ID, and
+  // if unable to do that returns the `Invalid` ID.
+  static auto TryMakeValue(int64_t value) -> IntId {
+    if (IntId::MinValue <= value && value <= IntId::MaxValue) {
+      return IntId(value);
+    }
+
+    return IntId::Invalid;
+  }
+
+  // Tries to make a signed APInt into an embedded value in the ID, and if
+  // unable to do that returns the `Invalid` ID.
+  static auto TryMakeSignedValue(llvm::APInt value) -> IntId {
+    if (value.sge(IntId::MinValue) && value.sle(IntId::MaxValue)) {
+      return IntId(value.getSExtValue());
+    }
+
+    return IntId::Invalid;
+  }
+
+  // Tries to make an unsigned APInt into an embedded value in the ID, and if
+  // unable to do that returns the `Invalid` ID.
+  static auto TryMakeUnsignedValue(llvm::APInt value) -> IntId {
+    if (value.ule(IntId::MaxValue)) {
+      return IntId(value.getZExtValue());
+    }
+
+    return IntId::Invalid;
+  }
+
+  // Pick a canonical bit width for the provided number of significant bits.
+  static auto CanonicalBitWidth(int significant_bits) -> int;
+
+  // Canonicalize an incoming signed APInt to the correct bit width.
+  static auto CanonicalizeSigned(llvm::APInt value) -> llvm::APInt;
+
+  // Canonicalize an incoming unsigned APInt to the correct bit width.
+  static auto CanonicalizeUnsigned(llvm::APInt value) -> llvm::APInt;
+
+  // Helper functions for handling values that are large enough to require an
+  // allocated `APInt` for storage. Creating or manipulating that storage is
+  // only a few lines of code, but it ends up expensive and a lot of code so we
+  // move these out-of-line.
```

Comment:
```suggestion
  // only a few lines of code, but we move these out-of-line because the generated code is big and harms performance for the non-`Large` common case.
```

Suggesting a slightly different comment due to discussion.